src/kern/npf_alg_pptp.c

```diff
@@ -0,0 +1,763 @@
+	if (ret)
+		return ret;
+
+	/* associate GRE ALG with the gre connection */
+	npf_nat_setalg(con->c_nat, pptp_alg.alg_pptp_gre, (uintptr_t)gre_con->u64);
+
+	NPF_DPRINTFCL(NPF_DC_PPTP_ALG, 50,
+			  "new pptp gre connection's nat %p\n", con->c_nat);
+
+	/* make gre connection active and pass */
+	npf_conn_set_active_pass(con);
+
+	gre_con->flags |= PPTP_ALG_FL_GRE_STATE_ESTABLISHED;
+	return 0;
+}
+
+static uint16_t
+npfa_translated_call_id_get(uint32_t ip)
+{
+	in_port_t port;
+	npf_addr_t addr;
+
+	addr.word32[0] = ip;
+	port = npf_portmap_get_pm(pptp_alg.pm, &pptp_alg.pm_params, sizeof(ip),
+			  &addr);
+
+	NPF_DPRINTFCL(NPF_DC_PPTP_ALG, 50,
+			  "pptp alg: get call_id %hu from "
+			  "the pormap ip %u\n", port, ip);
+
+	return (uint16_t)port;
+}
+
+static void
+npfa_translated_call_id_put(uint32_t ip, uint16_t call_id)
+{
+	npf_addr_t addr;
+	addr.word32[0] = ip;
+
+	npf_portmap_put_pm(pptp_alg.pm, sizeof(ip), &addr, (in_port_t)call_id);
+
+	NPF_DPRINTFCL(NPF_DC_PPTP_ALG, 50,
+			  "pptp alg: put call_id %hu to "
+			  "the pormap ip %u\n", call_id, ip);
+}
+
+/* lookup for and explicitly expire
+ * the associatiated gre connection state
+ */
+static void
+npfa_pptp_expire_gre_con(npf_t *npf, struct pptp_gre_con *gre_con,
+		  uint32_t client_ip, uint32_t server_ip)
+{
+	npf_conn_t *con;
+	npf_connkey_t key;
+	bool forw;
+
+	NPF_DPRINTFCL(NPF_DC_PPTP_ALG, 50,
+			  "expire gre con: orig client call_id %hu, client call_id %hu, "
+			  "server call_id %hu, flags %hu, cl %u, srv %u\n",
+			  gre_con->orig_client_call_id, gre_con->ctx.client_call_id,
+			  gre_con->ctx.server_call_id, gre_con->flags, client_ip, server_ip);
+
+	/* exit if gre connection was not established */
+	if ((gre_con->flags & PPTP_ALG_FL_GRE_STATE_ESTABLISHED) == 0)
+		return;
+
+	/* return translated call-id value back to the portmap */
+	if (gre_con->ctx.client_call_id != 0) {
+		npfa_translated_call_id_put(server_ip, gre_con->ctx.client_call_id);
+		NPF_DPRINTFCL(NPF_DC_PPTP_ALG, 50,
+				  "pptp control connection: put call_id %hu back to "
+				  "the pormap ip %u\n", gre_con->ctx.client_call_id, server_ip);
+	}
+
+	/* init a forward gre key */
+	npf_conn_init_ipv4_key((void *)&key, IPPROTO_GRE,
+			  gre_con->ctx.server_call_id, 0, client_ip, server_ip);
+
+	NPF_HEX_DUMPCL(NPF_DC_PPTP_ALG, 50, "gre key", &key,
+			  NPF_CONNKEY_V4WORDS * sizeof(uint32_t));
+
+	/* lookup for the associated pptp gre connection */
+	con = npf_conndb_lookup(npf->conn_db, &key, &forw);
+	if (con == NULL) {
+		NPF_DPRINTFCL(NPF_DC_PPTP_ALG, 50,
+			  "pptp control connection: associated gre conn not found, "
+			  "server call_id %hu\n", gre_con->ctx.server_call_id);
+		return;
+	}
+
+	NPF_DPRINTFCL(NPF_DC_PPTP_ALG, 50,
+			  "pptp control connection: expire associated gre conn "
+			  "server call_id %hu\n", gre_con->ctx.server_call_id);
+
+	/* mark the gre connection as expired */
+	npf_conn_set_expire(con);
+
+	gre_con->flags &= ~PPTP_ALG_FL_GRE_STATE_ESTABLISHED;
```

Comment:
Is this safe? As soon as you expire connection (call `npf_conn_expire()` above) -- if there are no other references -- another thread in paralle might destroy it (and call the ALG destructor too). So, this looks like a race condition.

Reply:
It's safe since "con" here is a gre connection and "gre_con" is a slot in the nat alg of the parent tcp conntrol  connection, and expiring gre connection doesn't affect parent tcp connection. I've renamed gre_con to gre_slot in the new code to avoid such confusion.